### ISSUE TITLE
MAgent EzPickle fail

### ISSUE THREAD

**Issue by dimikout3** [PettingZoo-Team/PettingZoo#345]
I am trying to use multiple clones of an environment by using the EzPickle library, as presented in the following code
```
env = battle_v2.parallel_env()
obs = env.reset()
_ = env.step(actions)
env2 = pickle.loads(pickle.dumps(env))
_ = env2.step(actions)
```
However, during the last line of code (execution of step in the cloned environment) I am getting the following error
```
Traceback (most recent call last):
  File "test.py", line 37, in <module>
    _ = env2.step(actions)
  File "/home/dkoutras/repos/MultiAgent/PettingZoo/pettingzoo/magent/magent_env.py", line 119, in step
    self.agents = [agent for agent in self.agents if not self.all_dones[agent]]
  File "/home/dkoutras/repos/MultiAgent/PettingZoo/pettingzoo/magent/magent_env.py", line 119, in <listcomp>
    self.agents = [agent for agent in self.agents if not self.all_dones[agent]]
AttributeError: '_parallel_env' object has no attribute 'all_dones'
```
Any ideas ?? I am pretty stuck with this one ...

**Comment by dimikout3** [comment]
I have also tried using the copy.deepcopy(env) library, yet I am getting the same error. It seems that some parameters from the original env are not copied/pickled, due to recursion reasons. I have tried to rewrite __deepcopy(self, memo)__ function in specific classes, yet still I failed ...

**Comment by weepingwillowben** [comment]
You need to reset the environment after loading it. The following code snippet works:

```
from pettingzoo.magent import battle_v2
import pickle
env = battle_v2.parallel_env()
obs = env.reset()
_ = env.step({agent:env.action_spaces[agent].sample() for agent in env.agents})
env2 = pickle.loads(pickle.dumps(env))
env2.reset()
_ = env2.step({agent:env.action_spaces[agent].sample() for agent in env.agents})
```

I'm sort of sorry that the EzPickle hack is hard to understand, but I don't really see the alternative.

**Comment by dimikout3** [comment]
Thanks for the fast response and the awesome work you are doing with the PettingZoo. I knew that I had to env.reset() the environment, but via this way I am "losing" any information that the original env had. I need something like cloning the original environment at a specific moment, in order to perform a bread first search (BFS). The ALE (Arcade Learning Environment) provides for example some functions like get_state(), ge_full_state(), set_state(), set_full_state(). @weepingwillowben  do you know if a similar functionality is available for pettingzoo ?

**Comment by dimikout3** [comment]
You can have a look at:
https://github.com/openai/gym/blob/master/gym/envs/atari/atari_env.py#L189

**Comment by weepingwillowben** [comment]
Unfortunately MAgent does not support this functionality. Adding it would require diving into the [c++ code](https://github.com/PettingZoo-Team/MAgent) and adding code to serialize all the variables in all of the objects. This was deemed to be too much effort. 

The raw Atari Learning Environment does have the c++ code needed to support serialization. However, the Atari Pettingzoo environment does not yet support this functionality. I am open to the idea of adding support for this. The Gym code you linked to can basically be copy-pasted into pettingzoo. However, the pettingzoo environment hides the implementation details of the base environment with wrappers. We can add gym's `.unwrapped` parameter to get around this. @justinkterry Is this something we want to support? I think so.

In the meantime, this functionality can be hacked in by doing something like:

```
from pettingzoo.atari import space_invaders_v1
env = space_invaders_v1.parallel_env()
ale = env.aec_env.env.env.env.ale # ideally this would probably just be env.unwrapped.ale

# taken from the gym code you linked to
state_ref = ale.cloneState()
state = ale.encodeState(state_ref)
ale.deleteState(state_ref)
```

In addition, pettingzoo environments are inspired by the real world. In theory, you should be able to make a pettingzoo environment in the real world, like was [made for gym](https://gym.offworld.ai/). It is impossible to save state in the real world. The whole field of reinforcement learning is based on the idea that you can only progress the MDP forward, not save or go backwards. So I am reluctant to add it to the API as anything other than an environment specific hack.

**Comment by justinkterry** [comment]
I agree that general API support isn't exactly optimal and that MAgent support is too much work. It seems like adding this for Atari may be worth doing, let's discuss this at the group meeting in a few hours.

**Comment by dimikout3** [comment]
> Unfortunately MAgent does not support this functionality. Adding it would require diving into the [c++ code](https://github.com/PettingZoo-Team/MAgent) and adding code to serialize all the variables in all of the objects. This was deemed to be too much effort.
> 
> The raw Atari Learning Environment does have the c++ code needed to support serialization. However, the Atari Pettingzoo environment does not yet support this functionality. I am open to the idea of adding support for this. The Gym code you linked to can basically be copy-pasted into pettingzoo. However, the pettingzoo environment hides the implementation details of the base environment with wrappers. We can add gym's `.unwrapped` parameter to get around this. @justinkterry Is this something we want to support? I think so.
> 
> In the meantime, this functionality can be hacked in by doing something like:
> 
> ```
> from pettingzoo.atari import space_invaders_v1
> env = space_invaders_v1.parallel_env()
> ale = env.aec_env.env.env.env.ale # ideally this would probably just be env.unwrapped.ale
> 
> # taken from the gym code you linked to
> state_ref = ale.cloneState()
> state = ale.encodeState(state_ref)
> ale.deleteState(state_ref)
> ```
> 
> In addition, pettingzoo environments are inspired by the real world. In theory, you should be able to make a pettingzoo environment in the real world, like was [made for gym](https://gym.offworld.ai/). It is impossible to save state in the real world. The whole field of reinforcement learning is based on the idea that you can only progress the MDP forward, not save or go backwards. So I am reluctant to add it to the API as anything other than an environment specific hack.

**Comment by justinkterry** [comment]
So we discussed this and we are going to enable this for Atari, but it'll probably be about a month before we're able to get aroudn to it

**Comment by justinkterry** [comment]
If you'd like to contribute that functionality to MAgent though (or Atari if you need it sooner), that'd be great.

**Comment by dimikout3** [comment]
@justinkterry great! thank you for your effort and time! We will have a look and let you know regarding our progress.

**Comment by weepingwillowben** [comment]
The relevant feature additions have been released in pettingzoo 1.6.1. You can now do

```
from pettingzoo.atari import space_invaders_v1
import time
env = space_invaders_v1.parallel_env()
env.reset()
for i in range(2000):
    env.step({agent:0 for agent in env.possible_agents})
state = env.unwrapped.clone_state()

env2 = space_invaders_v1.parallel_env()
env2.reset()
env2.unwrapped.restore_state(state) 
# unfortunately this is necessary to do before render. Not sure why...
env2.step({agent:0 for agent in env.possible_agents})
env2.render()
time.sleep(10)
```
